### PR TITLE
Azure terraform

### PR DESCRIPTION
**Pull request by gonzalogarciajaubert**
## Description
Synth, Deploy and Nuke using Terraform

## Changes

* Add new synth method: This method will generate a terraform template based on the user project
* Update deploy method: This method will deploy a terraform template based on the user project. The template is generated (synth) each time a deployment is run
* Update nuke method: This method will destroy all terraform resources

## Checks
- [X] Project Builds
- [X] Project passes tests and checks
- [x] Updated documentation accordingly => This will be done in a new PR
- [x] Add unit tests
- [x] Fix integration tests

### CODE REVIEW

File: packages/cli/src/commands/synth.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  synther: (config: BoosterConfig, logger: Logger) => Promise<void>
+): Promise<void> =>
+  Script.init(`boost ${Brand.dangerize('synth')} [${currentEnvironment()}] 🚀`, compileAndLoad)
+    .step('Synth', (config) => synther(config, logger))
+    .step('Cleaning up temporal files', cleanDeploymentSandbox)
+    .info('Synth complete!')
+    .done()
+
+export default class Synth extends BaseCommand {
+  public static description = 'Synth the current application as configured in your `index.ts` file.'
```
Comment (NickSeagull) [inline packages/cli/src/commands/synth.ts:26]: Maybe...
```suggestion
  public static description = 'Generate the required cloud templates to deploy your app manually.'
```

File: packages/framework-common-helpers/src/run-command.ts [hunk 2]
```diff
@@ -7,6 +7,7 @@ export function runCommand(
 ): ChildProcessPromise<PromiseResult<string>> {
   const subprocess = exec(command, {
     cwd: path,
+    maxBuffer: 1024 * 1024 * 1024,
```
Comment (NickSeagull) [inline packages/framework-common-helpers/src/run-command.ts:10]: Why is this needed and what are the implications?

File: packages/framework-provider-azure-infrastructure/src/infrastructure/helper/function-zip.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
+import { BoosterConfig } from '@boostercloud/framework-types'
+import { GraphqlFunction } from '../functions/graphql-function'
+import { EventHandlerFunction } from '../functions/event-handler-function'
+import { SchedulesFunctions } from '../functions/schedules-functions'
+import { FunctionDefinition } from '../types/functionDefinition'
+import archiver = require('archiver')
```
Comment (NickSeagull) [inline packages/framework-provider-azure-infrastructure/src/infrastructure/helper/function-zip.ts:6]: Why the `= require(...)`? Why not importing directly?

File: packages/framework-provider-azure-infrastructure/src/infrastructure/helper/function-zip.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    password: string,
+    functionAppName: string
+  ): Promise<any> {
+    needle.defaults({
+      open_timeout: 0,
+    })
+
+    return needle(
+      'post',
+      `https://${functionAppName}.scm.azurewebsites.net/api/zipDeploy`,
```
Comment (NickSeagull) [inline packages/framework-provider-azure-infrastructure/src/infrastructure/helper/function-zip.ts:58]: Maybe this should be a function that generates the name, so its easier to change if needed?

File: packages/framework-provider-azure-infrastructure/src/infrastructure/helper/function-zip.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        password: password,
+      }
+    )
+  }
+
+  private static async zipAzureFunction(config: BoosterConfig): Promise<any> {
+    const graphqlFunctionDefinition = new GraphqlFunction(config).getFunctionDefinition()
+    const eventHandlerFunctionDefinition = new EventHandlerFunction(config).getFunctionDefinition()
+    let featuresDefinitions = [graphqlFunctionDefinition, eventHandlerFunctionDefinition]
+    const schedulesFunctionsDefinition = new SchedulesFunctions(config).getFunctionDefinitions()
```
Comment (NickSeagull) [inline packages/framework-provider-azure-infrastructure/src/infrastructure/helper/function-zip.ts:71]: Typo:
```suggestion
    const scheduledFunctionsDefinition = new ScheduledFunctions(config).getFunctionDefinitions()
```

File: packages/framework-provider-azure-infrastructure/src/infrastructure/index.ts [hunk 6]
```diff
@@ -1,39 +1,76 @@
...
-    createWebSiteManagementClient(credentials),
-  ])
-  const resourceGroupName = createResourceGroupName(config)
-  await createResourceGroup(resourceGroupName, resourceManagementClient)
-  const applicationBuilder = new ApplicationStackBuilder(config)
-  await applicationBuilder.buildOn(logger, resourceManagementClient, webSiteManagementClient, resourceGroupName)
+  logger.info(`Deploying app ${config.appName}`)
+  const zipResource = await terraformSynth(logger, config)
+  const command = await runCommand(process.cwd(), 'cdktf deploy --auto-approve')
+  if (command?.stdout.includes('non-zero exit code')) {
```
Comment (NickSeagull) [inline packages/framework-provider-azure-infrastructure/src/infrastructure/index.ts:38]: Why not just checking the exit code? I believe its more standard

File: packages/framework-provider-azure-infrastructure/src/infrastructure/index.ts [hunk 7]
```diff
@@ -1,39 +1,76 @@
...
+
+  if (command?.stdout.includes('non-zero exit code')) {
+    return Promise.reject(`Destroy application ${config.appName} failed. Check cdktf logs`)
+  }
+}
+
+async function generateSynthFiles(logger: Logger, config: BoosterConfig): Promise<void> {
+  logger.info('Generating cdktf files')
+  const filesToGenerate: Array<[Array<string>, string]> = [[['cdktf.json'], ckdtfTemplate.template]]
+  await Promise.all(filesToGenerate.map(renderToFile(config)))
```
Comment (NickSeagull) [inline packages/framework-provider-azure-infrastructure/src/infrastructure/index.ts:58]: Note that `Promise.all` can fail silently, better use `Promise.allSettledAndFulfilled` (I believe there's a helper function somewhere)

File: packages/framework-provider-azure-infrastructure/src/infrastructure/synth/TerraformContainers.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      partitionKeyPath: `/${eventsStoreAttributes.partitionKey}`,
+      partitionKeyVersion: 2,
+      throughput: 400,
+    })
+
+    const readModels = Object.keys(config.readModels).map((readModel) =>
+      this.createReadModel(terraformStack, config, readModel, cosmosdbDatabase, cosmosdbSqlDatabase)
+    )
+
+    return [cosmosdbSqlEventContainer].concat(readModels)
```
Comment (NickSeagull) [inline packages/framework-provider-azure-infrastructure/src/infrastructure/synth/TerraformContainers.ts:31]: Maybe separate the creation of the event container into another function and make this `build` method only do the `concat`?

File: packages/framework-provider-azure-infrastructure/src/infrastructure/templates/cdktf.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+export const template = `{
+    "language": "typescript",
+    "app": "npm run --silent compile",
+    "projectId": "c13b5f64-7c69-4e12-93f9-a6daecd4f385",
```
Comment (NickSeagull) [inline packages/framework-provider-azure-infrastructure/src/infrastructure/templates/cdktf.ts:4]: Shouldn't the projectId be dynamic?

**Review by NickSeagull — CHANGES_REQUESTED** [review]
Left a couple of suggestions, but looking good! This rocks 😄

File: packages/cli/src/commands/synth.ts (hunk 1 above)
Comment (gonzalogarciajaubert) [reply packages/cli/src/commands/synth.ts:26]: Much better, thanks :)

File: packages/framework-provider-azure-infrastructure/src/infrastructure/helper/function-zip.ts (hunk 3 above)
Comment (gonzalogarciajaubert) [reply packages/framework-provider-azure-infrastructure/src/infrastructure/helper/function-zip.ts:6]: Ops! some wrong auto import. let me fix it

File: packages/framework-provider-azure-infrastructure/src/infrastructure/helper/function-zip.ts (hunk 4 above)
Comment (gonzalogarciajaubert) [reply packages/framework-provider-azure-infrastructure/src/infrastructure/helper/function-zip.ts:58]: Yep!

File: packages/framework-provider-azure-infrastructure/src/infrastructure/index.ts (hunk 6 above)
Comment (gonzalogarciajaubert) [reply packages/framework-provider-azure-infrastructure/src/infrastructure/index.ts:38]: yeah, let me fix it

File: packages/framework-provider-azure-infrastructure/src/infrastructure/index.ts (hunk 7 above)
Comment (gonzalogarciajaubert) [reply packages/framework-provider-azure-infrastructure/src/infrastructure/index.ts:58]: Awesome

File: packages/framework-provider-azure-infrastructure/src/infrastructure/synth/TerraformContainers.ts (hunk 8 above)
Comment (gonzalogarciajaubert) [reply packages/framework-provider-azure-infrastructure/src/infrastructure/synth/TerraformContainers.ts:31]: Yeah. I'm going to do something similar, let me change it and tell me what you think about it

File: packages/framework-common-helpers/src/run-command.ts (hunk 2 above)
Comment (gonzalogarciajaubert) [reply packages/framework-common-helpers/src/run-command.ts:10]: Cdktf logs the progress while deploying the application. The problem is caused by `cdktf` sending a lot of data (more than 1MB) to stdout or stderr: `Error: ChildProcessError: stdout maxBuffer length exceeded`

We need to increase the std buffer size

File: packages/framework-provider-azure-infrastructure/src/infrastructure/templates/cdktf.ts (hunk 9 above)
Comment (gonzalogarciajaubert) [reply packages/framework-provider-azure-infrastructure/src/infrastructure/templates/cdktf.ts:4]: I guess we don't need this parameter. Let me check it and remove it if possible

**Review by NickSeagull — APPROVED** [review]

File: packages/framework-provider-azure-infrastructure/src/infrastructure/synth/terraform-application-service-plan.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+import { toTerraformName } from '../helper/utils'
+
+export class TerraformApplicationServicePlan {
+  static build(
+    terraformStack: TerraformStack,
+    resourceGroup: ResourceGroup,
+    appPrefix: string,
+    resourceGroupName: string
+  ): AppServicePlan {
+    const id = toTerraformName(appPrefix, 'hpn')
```
Comment (alvaroloes) [inline packages/framework-provider-azure-infrastructure/src/infrastructure/synth/terraform-application-service-plan.ts:12]: What does `hpn` stands for?

File: packages/framework-provider-azure-infrastructure/src/infrastructure/synth/terraform-cosmosdb-sql-database.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    appPrefix: string,
+    cosmosdbAccount: CosmosdbAccount,
+    config: BoosterConfig
+  ): CosmosdbSqlDatabase {
+    const idDatabase = toTerraformName(appPrefix, 'dbd')
+    return new CosmosdbSqlDatabase(terraformStack, idDatabase, {
+      name: config.resourceNames.applicationStack,
+      resourceGroupName: resourceGroup.name,
+      accountName: cosmosdbAccount.name,
+      throughput: 400,
```
Comment (alvaroloes) [inline packages/framework-provider-azure-infrastructure/src/infrastructure/synth/terraform-cosmosdb-sql-database.ts:19]: Is there a way to configure the throughput of both the events store and the read models store to be automatic?
I mean, Is there an option for auto-scalability?

**Review by alvaroloes — APPROVED** [review]
Approved! left mainly questions

**Comment by alvaroloes** [comment]
@gonzalogarciajaubert Remember to execute integration tests and, when merging, change the title of the `Merge` commit to include a prefix like: `feat(azure): ...` so that a new version is released

**Comment by gonzalogarciajaubert** [comment]
@alvaroloes of course!. IT passed so I'm going to merge it

**Review by javiertoledo — COMMENTED** [review]
Very nice and clean code, great work @gonzalogarciajaubert!!

File: packages/framework-integration-tests/integration/helper/app-helper.ts [hunk 12]
```diff
@@ -12,9 +12,10 @@ export function applicationName(): string {
 
 export async function getProviderTestHelper(): Promise<ProviderTestHelper> {
   const provider = process.env.TESTED_PROVIDER
+  const environmentName = process.env.BOOSTER_ENV ?? 'azure'
```
[the remark this answers is not in the document]
Comment (gonzalogarciajaubert) [reply packages/framework-integration-tests/integration/helper/app-helper.ts:15]: You are right. Let me create a new ticket for that

File: packages/framework-integration-tests/integration/helper/app-helper.ts [hunk 13]
```diff
@@ -12,9 +12,10 @@ export function applicationName(): string {
...
   const providerHelpers: Record<string, () => Promise<ProviderTestHelper>> = {
     AWS: () => AWSTestHelper.build(applicationName()),
-    AZURE: () => AzureTestHelper.build(applicationName()),
+    AZURE: () => AzureTestHelper.build(applicationName(), environmentName),
```
[the remark this answers is not in the document]
Comment (gonzalogarciajaubert) [reply packages/framework-integration-tests/integration/helper/app-helper.ts:18]: You are right again :)

File: packages/framework-integration-tests/integration/provider-specific/azure/deployment/deployment.integration.ts [hunk 14]
```diff
@@ -8,7 +8,8 @@ describe('After deployment', () => {
       // The project must have been deployed by the cliHelper hook in setup.ts
       // that scripts uses the cli to do the deployment, so we just check here
       // that the resource group exists
-      await expect(AzureTestHelper.checkResourceGroup(applicationName())).to.be.eventually.fulfilled
+      const environmentName = process.env.BOOSTER_ENV ?? 'azure'
```
[the remark this answers is not in the document]
Comment (gonzalogarciajaubert) [reply packages/framework-integration-tests/integration/provider-specific/azure/deployment/deployment.integration.ts:11]: Yep

File: packages/framework-integration-tests/integration/provider-specific/azure/nuke/nuke.integration.ts [hunk 15]
```diff
@@ -5,9 +5,10 @@ import { expect } from '../../../helper/expect'
 describe('After nuke', () => {
   describe('the resource group', () => {
     it('is deleted successfully', async () => {
-      await expect(AzureTestHelper.checkResourceGroup(applicationName())).to.be.eventually.rejectedWith(
-        'ResourceGroupNotFound'
-      )
+      const environmentName = process.env.BOOSTER_ENV ?? 'azure'
```
[the remark this answers is not in the document]
Comment (gonzalogarciajaubert) [reply packages/framework-integration-tests/integration/provider-specific/azure/nuke/nuke.integration.ts:8]: LOL, sorry

File: packages/framework-integration-tests/integration/helper/app-helper.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (gonzalogarciajaubert) [reply packages/framework-integration-tests/integration/helper/app-helper.ts:15]: Ticket: https://agilemonkeys.atlassian.net/browse/BOOSTER-1652
